### DISCUSSION TITLE
error about pytorch.jit when run *.exe

### DISCUSSION THREAD

**Discussion by xiang-zhe** [pyinstaller/pyinstaller#6230] [Help]
win10
python3.8
pytorch == 1.9+cu11
torchvison == 0.10.0+cu11
cuda ==11.4
pyinstaller == 4.5
```
pytorch show：
torch.__version__: 1.9.0+cu111
torch.version.cuda: 11.1
torch.backends.cudnn.version: 8005
```

the PYINSTALLER works when code looks like that although some warnings about not found *.dll
````
import torch
from torchvision import transforms
import numpy as np
if __name__ == '__main__':
   bgr = torch.tensor([255, 0, 0], device="cuda").div(255).view( 1, 3, 1, 1)
   print(bgr)
````
but not work with my code after generating *.exe, my code works on my PC just running as `python ***.py`,
 ```
 File "runforzmq.py", line 115, in __init__
  File "torch\jit\_script.py", line 1096, in script
    return torch.jit._recursive.create_script_module(
  File "torch\jit\_recursive.py", line 411, in create_script_module
    AttributeTypeIsSupportedChecker().check(nn_module)
  File "torch\jit\_check.py", line 65, in check
    source_lines = textwrap.dedent(inspect.getsource(nn_module.__class__.__init__))
  File "inspect.py", line 997, in getsource
  File "inspect.py", line 979, in getsourcelines
  File "inspect.py", line 798, in findsource
OSError: could not get source code
```
it seem similar with https://github.com/pyinstaller/pyinstaller/issues/5673, but not the same, a lot infos reported about
```
File "torch/jit/_script.py", line 936, in script
  File "torch/jit/frontend.py", line 197, in get_jit_def
  File "torch/_utils_internal.py", line 56, in get_source_lines_and_file
OSError: Can't get source for <function c2p_dynamic_expand at 0x7ff34dfe7280>. TorchScript requires source access in order to carry out compilation, make sure original .py files are available.
[138475] Failed to execute script app
```
maybe they are same problem, but i didn't get it work,
1,  like:
```

def script_method(fn, _rcb=None):
    return fn
def script(obj, optimize=True, _frames_up=0, _rcb=None):
    return obj    
import torch.jit
torch.jit.script_method = script_method 
torch.jit.script = script


```
output
`RuntimeError: Freezing expects a ScriptModule as input. Please use torch.jit.script or torch.jit.trace to script your 'nn.Module'.`

2, change torchvision to 0.2.2.pos3, like :
`torchvision==0.2.2.pos3`

```
[5580] WARNING: file already exists but should not: C:\Users\AppData\Local\Temp\_MEI55802\torch\_C.cp38-win_amd64.pyd
Traceback (most recent call last):
  File "runforzmq.py", line 12, in <module>
  File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
  File "__init__.py", line 1, in <module>
  File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
  File "model.py", line 7, in <module>
  File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
  File "mobilenetv3.py", line 2, in <module>
ModuleNotFoundError: No module named 'torchvision.models.mobilenetv3'
[18084] Failed to execute script 'runforzmq' due to unhandled exception!
```

3, i also download transormers and change  the function  c2p_dynamic_expand with @torch.jit._script_if_tracing annotation.
i still get the same error
`OSError: could not get source code`
4, i remove torch.jit like:
```
        self.model.load_state_dict(torch.load(checkpoint, map_location=device))
        #self.model = torch.jit.script(self.model)
        #self.model = torch.jit.freeze(self.model)
        self.device = device
```
it works, but it works without jit, any help will be greatly appreciated~！！

**Comment by rokm** [comment]
Pytorch JIT requires source .py files. This makes it rather incompatible with PyInstaller, which by default byte-compiles the .py files and collects compiled .pyc files. 

However, our loader does support returning the source code from .py files that are collected into the application bundle (either copied manually, collected via .spec file, or collected via hooks). But it currently does not support doing so for the entry-point script, only for other modules and packages. 

So you'd need to move the code from the entry point script to a function in a module, and then call that from the entry-point script. And ensure that the module's .py file is collected, for example using [`--add-data`](https://pyinstaller.readthedocs.io/en/stable/spec-files.html#adding-files-to-the-bundle) command-line switch. Or, if you are using a 3rd party package that makes use of torch JIT, you can use `--collect-all <package>` or (in your .spec file or a custom hook) the `PyInstaller.utils.hooks.collect_data_files` function with `py_files=True˙.

But for what is worth, the sample you provided initially,
```
import torch
from torchvision import transforms
import numpy as np
if __name__ == '__main__':
   bgr = torch.tensor([255, 0, 0], device="cuda").div(255).view( 1, 3, 1, 1)
   print(bgr)
```
seems to work when frozen with PyInstaller without JIT issues.

**Reply by xiang-zhe** [reply]
HI, @rokm , i still didn't make it work,
i tried some case like:
1, i write a new script that didn't include jit as the entry-point script, it not work;

2, i write a new script that didn't include jit as the entry-point script,
and changed *.spec file like:

```
a = Analysis(['my scrpit.py'],
    pathex=['c:\\myfolder\\model', 'c:...\\python38\\Lib\\site-packages\\torch\\jit'],
    datas=['c:...\\python38\\Lib\\site-packages\\torch\\jit','jit'],...)
```

it not work; but when commenting jit, i can found the jit folder in _MEIXXX foder;

3, i write a new script that didn't include jit as a entry-point script, and copied the jit('\python38\Lib\site-packages\torch\jit') folder to my project folder and `import jit as torch_jit`,

```
a = Analysis(['my scrpit.py'],
    pathex=['c:\\myfolder\\model', ['c:my\\folder\\jit','jit']],
    datas=['c:my\\folder\\jit','jit'],...)
```

it not work, but when commenting jit, i can found the jit folder in _MEIXXX foder;
thx!

**Comment by QiFuChina** [comment]
I also encount the same error likes you, and I tried your solution 4 to remove jit however it works. But I am not sure whether cause new error for this program, at least it seems everything normal for now. 
By the way, have you found new solution about this error?

**Reply by xiang-zhe** [reply]
> I also encount the same error likes you, and I tried your solution 4 to remove jit however it works. But I am not sure whether cause new error for this program, at least it seems everything normal for now. By the way, have you found new solution about this error?

NO found  new solution, removing jit maybe just a little slower but no other errors

**Comment by Hommoner** [comment]
reference from https://blog.csdn.net/zhuguiqin1/article/details/124268815

modify inspect.py below
```
def getfile(object):
    """Work out which source or compiled file an object was defined in."""
    if ismodule(object):
        if getattr(object, '__file__', None):
            return object.__file__
        raise TypeError('{!r} is a built-in module'.format(object))
    if isclass(object):
        if hasattr(object, '__module__'):
            module = sys.modules.get(object.__module__)
            if getattr(module, '__file__', None):
                return module.__file__
        raise TypeError('{!r} is a built-in class'.format(object))
    if ismethod(object):
        object = object.__func__
    if isfunction(object):
        object = object.__code__
        # print what source file is not exist
        print('***lack of*** ',object)              # <========= modify here to print which file need to run source code when error occurs
    if istraceback(object):
        object = object.tb_frame
    if isframe(object):
        object = object.f_code
    if iscode(object):
        return object.co_filename
    raise TypeError('module, class, method, function, traceback, frame, or '
                    'code object was expected, got {}'.format(
                    type(object).__name__))
```

then build and run the app, when error occurs, you can see which module need source code
and put then in your exe path
I used to build pyinstaller by onedir , have not try onefile

my target is 
windows python 3.9.9
pytorch cudnn 8.3.2 cuda 11.3
detrctron2 and for Maskrcnn-R50 model convert to tensorrt model  and do inference,
anomalib  train and inference,
yolov5,yolov7 do inference
finally I get everything work by adding  source module : detectron2,anomalib,onnx,openvino,kornia,fvcore,timm,torchvision,torch
